plugins/BEdita/Core/src/Model/Table/UsersTable.php

```diff
@@ -216,13 +217,43 @@ public function login(EventInterface $event)
         $this->updateAll(
             [
                 'last_login' => $this->timestamp(),
+                'num_login_err' => 0,
             ],
             [
                 'id' => $identity->getIdentifier(),
             ]
         );
     }
 
+    /**
+     * Update fields on authentication failure.
+     *
+     * @param \Cake\Event\EventInterface $event Dispatched event.
+     * @return void
+     */
+    public function authFailure(EventInterface $event): void
+    {
+        /** @var \Cake\Http\ServerRequest|null $request */
+        $request = $event->getData('request');
+        if (
+            empty($request) ||
+            (string)$request->getData('grant_type') !== 'password' ||
+            !$request instanceof ServerRequest
+        ) {
```

Comment:
It would be enough

```suggestion
        if (
            !$request instanceof ServerRequest ||
            (string)$request->getData('grant_type') !== 'password'
        ) {
```
I suggest to place first the instance check to avoid a possible warning calling `getData()`.